Stat Tracking Database

Brayzure opened the pull request:
Postgres only for now.

Also comes with a fix for stacking status intervals.

Brayzure committed:
Preliminary Stat Reporting

- reportInterval of 0 disables reporting

Brayzure committed:
Lint Stuff

Brayzure committed:
Clear Client Intervals On Ready
- Previously, intervals were created and ignored, which led to multiple intervals if the client re-emitted the ready event (often)

ZeLarpMaster commented on lib/Client.js:
```diff
@@ -33,7 +38,26 @@ class Client {
+    }
+
+    resetIntervals() {
+        Object.keys(this.intervals).forEach((key) => {
```
Why iterate keys if you only actually need the values? You could just use [Object.values](https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Objets_globaux/Object/values) instead.

Brayzure replied:
Good catch, just habit I guess.

ZeLarpMaster commented on lib/Client.js:
```diff
@@ -122,7 +151,11 @@ class Client {
+        this.intervals.statusInterval = setInterval(this.rotatingStatus.bind(this), config.statusInterval);
+    
+        if(config.reportInterval) {
+            this.intervals.statInterval = setInterval(this.clientReporting.bind(this), config.reportInterval);
```
`config.statusInterval` --> `this.intervals.statusInterval`  
but then  
`config.reportInterval` --> `this.intervals.statInterval`

Please stick to one name for clarity (I'd prefer reportInterval because of `this.clientReporting`)

Brayzure replied:
I think that might have been from right when I started, before I chose a better name for it. I'll fix that.

ZeLarpMaster commented on lib/Client.js:
```diff
@@ -149,6 +182,23 @@ class Client {
+        this.client.shards.forEach((shard) => {
+            latency.push(shard.latency);
+        });
+        return _.mean(latency);
```
[Optional] this whole function could be `_.mean(this.client.shards.map(shard => shard.latency))` or some variation of it (like using a variable for the array of latency instead of all inline)

Brayzure replied:
Probably a good idea.

ZeLarpMaster commented on lib/Client.js:
```diff
@@ -149,6 +182,23 @@ class Client {
+        this.database.insertStat("userCount", this.client.users.size);
+        this.database.insertStat("memoryUsage", process.memoryUsage().heapUsed);
+        this.database.insertStat("databaseLatency", Math.round(_.mean(this.database.latestLatencies)));
+        this.database.insertStat("shardLatency", this.getAverageShardLatency());
```
[Optional] I don't know if we'd ever need those stat type names elsewhere, but if we did, it would be great if those were in constants or something. Not really necessary for now since we don't, but might be worth actively adding them as constants instead of moving them the day we do need them.

Brayzure replied:
The way I see it, we should probably stick to one source of data for each stat type. If it's going elsewhere, then it will probably end up getting removed from here.

ZeLarpMaster commented on lib/core/db/PostgreSQLQueries.js:
```diff
@@ -29,6 +29,12 @@ queries.createTables = [
         + " ); ",
+    " CREATE TABLE IF NOT EXISTS stats "
+        + " ( "
+        + "     time timestamp with time zone, "
```
with timezone? Do we want it to be zoned or should we not zone it and always make it UTC (or even just a bigint like the `sent_at` field in emotes)?

ZeLarpMaster replied:
Also I'd like a more significative name on this. "time" is a bit too generic. Maybe something like "created_at" or "recorded_at" (don't forget to edit the query if you change this).

Brayzure replied:
If I understand it correctly, `with time zone` stores timezone along with the date and time, which allows accurate date storage regardless of where the bot is hosted. Caring about which time zone is used is something we may want to do outside of the database.

Brayzure replied:
`created_at` is a better name for it, let's use that.

Brayzure committed:
Requested Changes
- Object.values over Object.keys
- reportInterval over statInterval
- Shorter getAverageShardLatency
- created_at over time
